File: Ds.LaborManagement/Dominion.LaborManagement.Dto/Scheduling/ClockEmployeePunchDto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Dominion.Core.Dto.Labor;

namespace Dominion.LaborManagement.Dto.Scheduling
{
    [Serializable]
    public partial class ClockEmployeePunchDto : IHasPunchShiftDateInfo
    {
        public int ClockEmployeePunchId { get; set; }
        public int EmployeeId { get; set; }
        public string EmployeeFirstName { get; set; }
        public string EmployeeLastName { get; set; }
        public DateTime ModifiedPunch { get; set; }
        public DateTime RawPunch { get; set; }
        public int? ClientCostCenterId { get; set; }
        public int? ClientDivisionId { get; set; }
        public int? ClientDepartmentId { get; set; }
        public int? ClientShiftId { get; set; }
        public int RawPunchBy { get; set; }
        public int? ClockEmployeePunchTypeId { get; set; }
        public int? ClockClientLunchId { get; set; }
        public bool IsPaid { get; set; }
        public string Comment { get; set; }
        public DateTime? ShiftDate { get; set; }
        public bool? IsManualShiftOverride { get; set; }
        public byte? TimeZoneId { get; set; }
        public string ClockName { get; set; }
        public byte? TransferOption { get; set; }
        public bool IsStopAutoLunch { get; set; }
        public int ClientId { get; set; }
        public int? ClockEmployeeScheduleId { get; set; }
        public byte? ScheduleNumber { get; set; }
        public int? ClientJobCostingAssignmentId1 { get; set; }
        public int? ClientJobCostingAssignmentId2 { get; set; }
        public int? ClientJobCostingAssignmentId3 { get; set; }
        public int? ClientJobCostingAssignmentId4 { get; set; }
        public int? ClientJobCostingAssignmentId5 { get; set; }
        public int? ClientJobCostingAssignmentId6 { get; set; }
        public string EmployeeComment { get; set; }
        public int? ClockEmployeePunchLocationId { get; set; }
        public decimal? ClockEmployeePunchLocationLat { get; set; }
        public decimal? ClockEmployeePunchLocationLng { get; set; }

    }
}

File: WebCore/lib/ds/core/src/lib/employee-services/models/clock-employee-punch-dto.ts
import { Moment } from 'moment';
import { convertToMoment } from '@ds/core/shared/convert-to-moment.func';
import { ClockExceptionEnum } from '@ajs/labor/models/client-exception-detail.model';

export interface ClockEmployeePunchDto {
    clockEmployeePunchId: number;
    employeeId: number;
    employeeFirstName: string;
    employeeLastName: string;
    modifiedPunch: Moment|Date|string;
    rawPunch: Moment|Date|string;
    clientCostCenterId?: number;
    clientDivisionId?: number;
    clientDepartmentId?: number;
    clientShiftId?: number;
    rawPunchBy: number;
    clockEmployeePunchTypeId?: number;
    clockClientLunchId?: number;
    isPaid: boolean;
    comment: string;
    shiftDate?: Moment|Date|string;
    isManualShiftOverride?: boolean;
    timeZoneId?: number;
    clockName: string;
    transferOption?: number;
    isStopAutoLunch: boolean;
    clientId: number;
    clockEmployeeScheduleId?: number;
    scheduleNumber?: number;
    clientJobCostingAssignmentId1?: number;
    clientJobCostingAssignmentId2?: number;
    clientJobCostingAssignmentId3?: number;
    clientJobCostingAssignmentId4?: number;
    clientJobCostingAssignmentId5?: number;
    clientJobCostingAssignmentId6?: number;
    employeeComment: string;
    errorType: ClockExceptionEnum | null;
    // compare(that: ClockEmployeePunchDto): number;
    clockEmployeePunchLocationLat?: number;
    clockEmployeePunchLocationLng?: number;
    clockEmployeePunchLocationId: number | null;
    isChecked: boolean | null;
    isSelectable?: boolean;
}

export class ClockEmployeePunchDtoImpl implements ClockEmployeePunchDto {
    clockEmployeePunchId: number;
    employeeId: number;
    employeeFirstName: string;
    employeeLastName: string;
    modifiedPunch: Moment|Date|string;
    rawPunch: Moment|Date|string;
    clientCostCenterId?: number;
    clientDivisionId?: number;
    clientDepartmentId?: number;
    clientShiftId?: number;
    rawPunchBy: number;
    clockEmployeePunchTypeId?: number;
    clockClientLunchId?: number;
    isPaid: boolean;
    comment: string;
    shiftDate?: Moment|Date|string;
    isManualShiftOverride?: boolean;
    timeZoneId?: number;
    clockName: string;
    transferOption?: number;
    isStopAutoLunch: boolean;
    clientId: number;
    clockEmployeeScheduleId?: number;
    scheduleNumber?: number;
    clientJobCostingAssignmentId1?: number;
    clientJobCostingAssignmentId2?: number;
    clientJobCostingAssignmentId3?: number;
    clientJobCostingAssignmentId4?: number;
    clientJobCostingAssignmentId5?: number;
    clientJobCostingAssignmentId6?: number;
    employeeComment: string;
    errorType: ClockExceptionEnum | null;
    clockEmployeePunchLocationLat?: number;
    clockEmployeePunchLocationLng?: number;
    clockEmployeePunchLocationId: number | null;
    isChecked: boolean | null;
    isSelectable?: boolean;

    constructor(that: ClockEmployeePunchDto) {
        Object.getOwnPropertyNames(that).forEach(property => {
            this[property] = that[property];
        });
    }

    static punchExists(punch: ClockEmployeePunchDto): boolean {
      return typeof(punch) !== 'undefined';
    }

    static coalesceShiftDate(p: ClockEmployeePunchDto): Moment|Date|string {
        // Backend uses:
        // FilterBy(x => (x.ShiftDate ?? x.ModifiedPunch) >= startDate && (x.ShiftDate ?? x.ModifiedPunch) <= endDate);
        return p.shiftDate ? p.shiftDate : p.modifiedPunch;
    }

    // static comparePunches(a: Moment, b: Moment): number {
    static comparePunches(a: Moment|Date|string, b: Moment|Date|string): number {
        const aMoment = convertToMoment(a);
        const bMoment = convertToMoment(b);
        let result = 0;

        if (aMoment.isBefore(bMoment)) {
            result = -1;
        } else if (!aMoment.isSame(bMoment)) {
            result = 1;
        }

        return result;
    }

    compare(that: ClockEmployeePunchDto): number {
        return ClockEmployeePunchDtoImpl.comparePunches(this.rawPunch, that.rawPunch);
    }
}
